ci/docker.go

```diff
@@ -31,38 +31,46 @@ func (d *Docker) Run(ctx context.Context, job *Job, user string, timeout time.Du
 		return "", err
 	}
 
-	if err := pullImage(ctx, cli, job.Image); err != nil {
-		return "", err
+	create := func() (container.ContainerCreateCreatedBody, error) {
+		return cli.ContainerCreate(ctx, &container.Config{
+			Image: job.Image,
+			Cmd:   []string{"/bin/bash", "-c", strings.Join(job.Commands, "\n")},
+		}, nil, nil, user)
 	}
 
-	resp, err := cli.ContainerCreate(ctx, &container.Config{
-		Image: job.Image,
-		Cmd:   []string{"/bin/sh", "-c", strings.Join(job.Commands, "\n")},
-	}, nil, nil, user)
+	resp, err := create()
 	if err != nil {
-		return "", err
+		// if image not found locally, try to pull it
+		if err := pullImage(ctx, cli, job.Image); err != nil {
+			return "", err
+		}
+		resp, err = create()
+		if err != nil {
+			return "", err
+		}
 	}
 
-	if csErr := cli.ContainerStart(ctx, resp.ID, types.ContainerStartOptions{}); csErr != nil {
-		return "", csErr
+	if err := cli.ContainerStart(ctx, resp.ID, types.ContainerStartOptions{}); err != nil {
+		return "", err
 	}
 
 	// will wait until the container stops
-	waitc, errc := cli.ContainerWait(ctx, resp.ID, container.WaitConditionNotRunning)
+	waitCh, errCh := cli.ContainerWait(ctx, resp.ID, container.WaitConditionNotRunning)
 
 	if timeout < 1 {
 		timeout = containerTimeout
 	}
 
 	select {
-	case wErr := <-errc:
-		fmt.Println("wErr: ", wErr.Error())
-		return "", wErr
-		// if the container still running after predefined time interval, force kill it
+	case err := <-errCh:
+		// container failed with error; return
+		return "", err
 	case <-time.After(timeout):
-		cli.ContainerKill(ctx, resp.ID, "SIGTERM")
-		return fmt.Sprintf("Container timed out after %v", timeout), nil
-	case <-waitc:
+		// force kill container after predefined time interval
+		err = cli.ContainerKill(ctx, resp.ID, "SIGTERM")
+		return "", fmt.Errorf("container timed out after %v: %w", timeout, err)
```

Comment:
Is it really wise to return the error here? This way there is no container output and the user receives no feedback about what has happened. 
Now, instead of receiving a hint about the timeout in the build info on the `Results` page, students will be left guessing what went wrong, because from their perspective it would look like the tests had never run and something is broken on our side.
Could it be better to change the output message to something more helpful, like hinting on possible presence of an infinite loop in the submitted code, instead of just failing?

Reply:
Valid point. Will fix. But then we can't return an error as well. It would be nice to also log this fact on the server side. Will look into it.